app/models/concerns/submission_processable.rb

```diff
@@ -0,0 +1,103 @@
+module SubmissionProcessable
+  extend ActiveSupport::Concern
+
+  private
+
+  def process_next_steps(data)
+    return if data.to_s.eql?('INTERNAL_SERVER_ERROR')
+
+    next_links = extract_next_links(data)
+    loop_each next_links if next_links.any?
+  end
+
+  def request_and_extract_data(uri)
+    parsed_uri = build_uri(uri)
+    data = request_endpoint(parsed_uri.for_calling)
+    extract_data_from(data, parsed_uri)
+    data
+  end
+
+  def loop_each(array_of_urls)
+    array_of_urls.each do |uri|
+      data = request_and_extract_data(uri)
+      process_next_steps(data)
+    end
+  end
+
+  def extract_data_from(data, parsed_uri)
+    if data.to_s.eql?('INTERNAL_SERVER_ERROR')
+      record_error_data(parsed_uri)
+    elsif data_is_valid?(data)
+      record_valid_data(data, parsed_uri)
+    end
+  end
+
+  def data_is_valid?(data)
+    (data.is_a?(Array) || data.is_a?(Hash)) && data.except('_links').keys.present?
+  end
+
+  def record_valid_data(data, parsed_uri)
+    key = build_data_key(data, parsed_uri)
+    value = build_data_value(data)
+    @result[:data] << { "#{key}": value }
+  end
+
+  def record_error_data(parsed_uri)
+    @result[:data] << { "#{parsed_uri.for_displaying}": 'returned INTERNAL_SERVER_ERROR' }
+  end
+
+  def build_data_value(data)
+    if data.is_a?(Hash)
+      data[data.except('_links').keys.first]
+    else
+      # :nocov:
+      # is this required? (it doesn't seem to be getting called,
+      # I am not sure what sort of data structure it would apply to?)
+      data[data.except('_links').keys.first].first
+      # :nocov:
```

Comment:
I'm not entirely sure I understand that line either - from the `data_is_valid?` method it looks like it's possible for `data` to be an array instead of a hash. I guess you could write a test to handle that scenario, although I don't know how likely it is. 

Colin probably knows more so I think you're probably safe to merge this as is and make changes when he's back if necessary.